salt.returners.pgjsonb: update jids row on conflict

baby-gnu opened the pull request:
### What does this PR do?

* salt/returners/pgjsonb.py: add “ON CONFLICT” clause to the “INSERT”
  to update the “load” column when the jid is already registered.
  Use named placeholders in SQL statement.

### What issues does this PR fix or reference?

This commit fix #47327.

### Previous Behavior
I had errors in PostgreSQL logs

```
ERROR:  duplicate key value violates unique constraint "jids_pkey"
DETAIL:  Key (jid)=(20180426141333546550) already exists.
STATEMENT:  INSERT INTO jids (jid, load) VALUES ('20180426141333546550', '{"fun_args": [], "jid": "20180426141333546550", "return": true, "retcode": 0, "success": true, "cmd": "_return", "_stamp": "2018-04-26T14:13:33.758033", "fun": "test.ping", "id": "5"}')
```

### New Behavior
I have no errors in PostgreSQL logs.

### Tests written?

No, I don't know how to test this.

### Commits signed with GPG?

Yes

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

mattp- commented:
this raises the minimum PG version to 9.5; when ON CONFLICT DO UPDATE was introduced. unsure if this an issue?
given its a single query, it would be nice if you could test the capability of the server and act accordingly.

cro commented:
For reference, from the Pg website at https://www.postgresql.org/support/versioning/

```
9.2 is EOL
9.3 EOL September 2018
9.4 EOL December 2019
9.5 EOL January 2021
```

cachedout commented:
@baby-gnu I think that @mattp- makes a very good point here. What do you think about that idea?

mattp- commented:
@baby-gnu looks like you would just have to do something like this
https://github.com/psycopg/psycopg2/blob/master/tests/test_connection.py#L90
since there is no object in a returner, you might want to create a PG_VERSION constant at the top, and inside _get_serv set PG_VERSION to conn.server_version if it is None

rallytime commented:
@mattp- How does this look to you now?

mattp- commented:
its going to be making the version check on each insert, rather than on first connection; I'm guessing it's cached locally by libpq though so it doesn't really make much of a difference. seems fine to me

baby-gnu commented:
> its going to be making the version check on each insert, rather than on first connection; I'm guessing it's cached locally by libpq though so it doesn't really make much of a difference. seems fine to me

I can rework it to make the check and define the SQL snippet only once if you prefer.

mattp- commented:
@baby-gnu sure :)

rallytime commented:
Hi @baby-gnu - Just checking in here to see if you've had a chance to implement the latest suggestions.

baby-gnu committed:
salt.returners.pgjsonb: update jids row on conflict

* salt/returners/pgjsonb.py: add “ON CONFLICT” clause to the “INSERT”
  to update the “load” column when the jid is already registered for
  supported PostgreSQL version.
  Use named placeholders in SQL statement.

baby-gnu committed:
salt.returners.pgjsonb: the “ON CONFLICT/DO UPDATE” was always used

* salt/returners/pgjsonb.py: fix the logic to match only version >= 9.5.

rallytime commented:
@mattp- How does this look to you now?

mattp- commented:
:+1: looks good

Commit:
Lint: Add blank line
